src/hotspot/share/gc/g1/g1HeapSizingPolicy.cpp

```diff
@@ -436,3 +446,262 @@ size_t G1HeapSizingPolicy::full_collection_resize_amount(bool& expand, size_t al
+
+  // Find time-based candidates
+  find_uncommit_candidates_by_time(&candidates, max_regions_to_shrink);
+
+  if (candidates.length() == 0) {
+    log_debug(gc, sizing)("Time-based shrink: no candidates found");
+    return 0;
+  }
+
+  uint valid_candidates = 0;
+  // Count valid candidates (must be available, empty, and free)
+  for (int i = 0; i < candidates.length(); i++) {
+    G1HeapRegion* hr = candidates.at(i);
+    uint region_index = hr->hrm_index();
+
+    // Only count if region is ready for shrinking
+    if (hr->is_empty() && hr->is_free()) {
+      log_debug(gc, sizing)("Time-based shrink: identified region %u as candidate (last_access=" UINT64_FORMAT "ms ago)",
+                           region_index, (Ticks::now() - hr->last_access_time()).milliseconds());
+      valid_candidates++;
+    } else {
+      log_debug(gc, sizing)("Time-based shrink: skipping region %u - not ready for shrinking "
+                           "(empty=%s, free=%s)",
+                           region_index, hr->is_empty() ? "true" : "false", hr->is_free() ? "true" : "false");
+    }
+  }
+
+  size_t shrink_bytes = (size_t)valid_candidates * G1HeapRegion::GrainBytes;
+
+  if (valid_candidates > 0) {
+    log_info(gc, sizing)("Time-based shrink: requesting %zuMB based on %u time-based candidates",
+                         shrink_bytes / M, valid_candidates);
+  }
+
+  return shrink_bytes;
+}
+
+bool G1HeapSizingPolicy::should_uncommit_region(G1HeapRegion* hr) const {
+  // Note: Caller already guarantees hr->is_empty() is true
+  // Empty regions should always be free and not in collection set in normal operation
+
+  Ticks current_time = Ticks::now();
+  Ticks last_access = hr->last_access_time();
+  Tickspan elapsed = current_time - last_access;
+
+  log_trace(gc, sizing)("Region %u uncommit check: elapsed=" JLONG_FORMAT "ms threshold=" JLONG_FORMAT "ms last_access=" JLONG_FORMAT " now=" JLONG_FORMAT " empty=%s",
+                     hr->hrm_index(), (jlong)elapsed.milliseconds(), (jlong)G1UncommitDelayMillis, last_access.value(), current_time.value(),
+                     hr->is_empty() ? "true" : "false");
+
+  bool should_uncommit = elapsed.milliseconds() > G1UncommitDelayMillis;
+  if (should_uncommit) {
+    log_debug(gc, sizing)("Region state transition: transitioning from active to idle");
+    log_debug(gc, sizing)("Region state transition: Region %u transitioning from active to idle after " JLONG_FORMAT "ms idle",
+                  hr->hrm_index(), (jlong)elapsed.milliseconds());
+  }
+
+  return should_uncommit;
+}
+
+size_t G1HeapSizingPolicy::evaluate_heap_resize_for_uncommit() {
+  if (!G1UseTimeBasedHeapSizing) {
+    return 0;
+  }
+
+  // Back off during allocation pressure - only evaluate when truly idle
+  if (_analytics != nullptr) {
+    double gc_time_ratio = _analytics->short_term_gc_time_ratio();
+    if (gc_time_ratio > 0.05) { // 5% GC time still indicates pressure
```

Comment:
gc time ratio is no indicator for idleness. Also the default is 4% right now (ie. which G1 aims for), this value would be higher than the goal. Also the time ratio is user-settable - some users may accept a higher gc usage than others.

I understand the need for this change to kind of see if there is "pressure" - however this would be implicit by gcs occurring inbetween (i.e. if the `ServiceTask` did not occur/did not find need for uncommitting regions between GCs due to time stamps too high, then there is nothing to do anyway).